cirq/contrib/two_qubit_gates/gate_compilation.py

```diff
@@ -0,0 +1,379 @@
+"""Attempt to tabulate single qubit gates required to generate a target 2Q gate
+with a product A k A."""
+from functools import reduce
+from typing import Tuple, Sequence, List, NamedTuple
+
+import numpy as np
+import attr
+from cirq import kak_decomposition, kak_vector, value
+from cirq.contrib.two_qubit_gates.math_utils import (kak_vector_infidelity,
+                                                     vector_kron,
+                                                     weyl_chamber_mesh,
+                                                     random_qubit_unitary,
+                                                     kak_vector_to_unitary)
+
+_SingleQubitGatePair = Tuple[np.ndarray, np.ndarray]
+
+
+class TwoQubitGateCompilation(NamedTuple):
+    """Represents a compilation of a target 2-qubit with respect to a base gate.
+
+    This object encodes the relationship between 4x4 unitary operators
+
+    U_target ~ k_N · U_base · k_{N-1} · ... · k_1 · U_base · k_0
+
+    where U_target, U_base are 2-local and k_j are 1-local.
+
+    Attributes:
+        base_gate: 4x4 unitary denoting U_base above.
+        target_gate: 4x4 unitary denoting U_target above.
+        local_unitaries: Sequence of 2-tuples (k_{00},k_{01}),(k_{10},k_{11})...
+            where k_j = k_{j0} ⊗ k_{j1} in the product above. Each k_{j0},
+            k_{j1} is a 2x2 unitary.
+        actual_gate: 4x4 unitary denoting the right hand side above, ideally
+            equal to U_target.
+        success: Whether actual_gate is expected to be close to U_target.
+    """
+    base_gate: np.ndarray
+    target_gate: np.ndarray
+    local_unitaries: Tuple[_SingleQubitGatePair, ...]
+    actual_gate: np.ndarray
+    success: bool
+
+
+@attr.s(auto_attribs=True)
+class GateTabulation:
+    """A 2-qubit gate compiler based on precomputing/tabulating gate products.
+
+    """
+    base_gate: np.ndarray  # Base two qubit gate. (4x4 unitary)
```

Comment:
Rename to `base_gate_unitary` to disambiguate w.r.t. `cirq.Gate` instances

Reply:
Done